## Establish Dependencies and Load in Data

In [1]:
# Import all packages and dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
import sys, warnings, os
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [2]:
#Ignore warnings when validating scores
if not sys.warnoptions:
    warnings.simplefilter("ignore")
    os.environ["PYTHONWARNINGS"] = "ignore"

In [3]:
#Setting columns and rows to display all the results
pd.set_option("display.max_columns", None, "display.max_rows", None)

In [4]:
# Read in the dataset
ufc_master_df = pd.read_csv("../Resources/ufc-master.csv")
ufc_master_df.head()

R_fighter         B_fighter  R_odds  B_odds        R_ev  \
0   Alistair Overeem  Alexander Volkov     150    -182  150.000000   
1     Cory Sandhagen     Frankie Edgar    -400     300   25.000000   
2  Alexandre Pantoja        Manel Kape    -125     100   80.000000   
3     Diego Ferreira    Beneil Dariush    -125     105   80.000000   
4    Michael Johnson        Clay Guida    -225     175   44.444444   

         B_ev      date                location country Winner  title_bout  \
0   54.945055  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
1  300.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
2  100.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
3  105.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
4  175.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   

   weight_class gender  no_of_rounds  B_current_lose_streak  \
0   Heavyweight   MALE             5                      0   
1  Bantamweight   MALE             3                      0   
2     Flyweight   MALE             3                      0   
3   Lightweight   MALE             3                      0   
4   Lightweight   MALE             3                      2   

   B_current_win_streak  B_draw  B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  \
0                     1       0                  4.76               0.58   
1                     1       1                  3.70               0.39   
2                     0       0                  0.00               0.00   
3                     5       1                  3.98               0.49   
4                     0       0                  2.41               0.33   

   B_avg_SUB_ATT  B_avg_TD_landed  B_avg_TD_pct  B_longest_win_streak  \
0            0.2             0.69          0.70                     4   
1            0.3             2.28          0.31                     5   
2            0.0             0.00          0.00                     0   
3            1.1             1.75          0.32                     5   
4            0.7             3.37          0.38                     4   

   B_losses  B_total_rounds_fought  B_total_title_bouts  \
0         2                     26                    0   
1         8                     94                    9   
2         0                      0                    0   
3         4                     36                    0   
4        15                     87                    0   

   B_win_by_Decision_Majority  B_win_by_Decision_Split  \
0                           0                        1   
1                           0                        1   
2                           0                        0   
3                           0                        1   
4                           0                        3   

   B_win_by_Decision_Unanimous  B_win_by_KO/TKO  B_win_by_Submission  \
0                            2                3                    0   
1                           10                5                    2   
2                            0                0                    0   
3                            4                3                    5   
4                            8                2                    4   

   B_win_by_TKO_Doctor_Stoppage  B_wins  B_Stance  B_Height_cms  B_Reach_cms  \
0                             0       6  Orthodox        200.66       203.20   
1                             0      18  Orthodox        167.64       172.72   
2                             0       0  Southpaw        165.10       172.72   
3                             0      13  Southpaw        177.80       182.88   
4                             0      17  Orthodox        170.18       177.80   

   B_Weight_lbs  R_current_lose_streak  R_current_win_streak  R_draw  \
0           250                      0                     2       0   
1           135                      0                     1       0   
2           

In [5]:
print(ufc_master_df.shape)
print(ufc_master_df.info())

(4566, 137)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4566 entries, 0 to 4565
Columns: 137 entries, R_fighter to B_rev_bout
dtypes: bool(1), float64(73), int64(47), object(16)
memory usage: 4.7+ MB
None


## Baseline Classification Model

In [6]:
# Splitting X (features) from y (labels)
#Classification task, the target value will be "Winner"
y = ufc_master_df.Winner
X = ufc_master_df.drop(['Winner'], axis=1)

In [7]:
# Begin encoding categorical variables by separating variabes by datatype
categorical_col = [col for col in X.columns if X[col].dtypes == 'object']
numeric_col = [col for col in X.columns if col not in categorical_col]

In [8]:
# Encode categorical columns first in order to avoid data leakage in split
enc = LabelEncoder()
for i in X[categorical_col]:
    #using astype(str) to avoid columns with 'float and str' to throw errors
    X[i] = enc.fit_transform(X[i].astype(str))
    

In [9]:
# Encode label (y) for Red fighter ID = 1 Blue fighter ID = 0
y = [1 if win == "Red" else 0 for win in y]
y

[0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,


In [10]:
# Split dataset into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size = 0.25)

In [11]:
# Find and fill null values
# List column key/values in descending order
X_train.isnull().sum().sort_values(ascending=False)

B_Women's Featherweight_rank    3424
R_Women's Featherweight_rank    3417
B_Pound-for-Pound_rank          3403
B_Women's Flyweight_rank        3396
R_Women's Flyweight_rank        3387
B_Women's Strawweight_rank      3386
B_Featherweight_rank            3365
B_Women's Bantamweight_rank     3365
B_Lightweight_rank              3363
B_Light Heavyweight_rank        3362
B_Bantamweight_rank             3362
B_Flyweight_rank                3358
B_Welterweight_rank             3357
B_Middleweight_rank             3356
R_Women's Strawweight_rank      3356
B_Heavyweight_rank              3346
R_Women's Bantamweight_rank     3343
R_Featherweight_rank            3333
R_Lightweight_rank              3329
R_Middleweight_rank             3329
R_Flyweight_rank                3328
R_Light Heavyweight_rank        3326
R_Bantamweight_rank             3323
R_Welterweight_rank             3317
R_Heavyweight_rank              3317
R_Pound-for-Pound_rank          3312
R_rev_bout                      1330
B

In [12]:
# Creating a StandardScaler instance.

scaler = StandardScaler()
# Fitting the Standard Scaler with the training data.
X_scaler = scaler.fit(X_train)

##Scaling the data.
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
# Saving impute scaler instances in case we decide to use instead of standar scaler
# Impute the numerical columns with mean.

#imp = SimpleImputer(strategy='most_frequent')
#scaler = imp.fit(X_train[numeric_col])
#X_train[numeric_col] = scaler.transform(X_train[numeric_col])
#X_test[numeric_col] = scaler.transform(X_test[numeric_col])

In [14]:
# Fill the categorical columns with 'most_frequent'
#imp1 = SimpleImputer(strategy='most_frequent')
#scaler1 = imp1.fit(X_train[categorical_col])
#X_train[categorical_col] = scaler1.transform(X_train[categorical_col])
#X_test[categorical_col] = scaler1.transform(X_test[categorical_col])

In [15]:
X_train.sample(10)

R_fighter  B_fighter  R_odds  B_odds        R_ev        B_ev  date  \
1763       1241       1054     375    -470  375.000000   21.276596   245   
1759        500        185    -380     315   26.315789  315.000000   282   
1506        369        782     150    -170  150.000000   58.823529   117   
2036        747         24     150    -170  150.000000   58.823529    56   
2544        331       1274    -380     315   26.315789  315.000000   364   
4394        962        961    -230     190   43.478261  190.000000    63   
2954       1099       1477     195    -215  195.000000   46.511628    41   
715         779        227    -115    -105   86.956522   95.238095    74   
2681        228       1457    -420     335   23.809524  335.000000   239   
4330        352       1139    -275     235   36.363636  235.000000   165   

      location  country  title_bout  weight_class  gender  no_of_rounds  \
1763       131       20       False             5       1             3   
1759       109        2       False             0       1             3   
1506        39       21       False             6       1             3   
2036        74       23       False             8       1             3   
2544        85       21       False             7       1             3   
4394        93        9       False             8       1             3   
2954       108        2       False             3       1             3   
715         88       21       False             8       1             3   
2681        66       21        True             7       1             5   
4330        66       21       False             6       1             3   

      B_current_lose_streak  B_current_win_streak  B_draw  \
1763                      0                     4       0   
1759                      0                     0       0   
1506                      0                     1       0   
2036                      1                     0       0   
2544                      0                     3       0   
4394                      0                     0       0   
2954                      0                     1       0   
715                       0                     2       0   
2681                      0                     3       0   
4330                      0                     1       0   

      B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  B_avg_SUB_ATT  B_avg_TD_landed  \
1763             27.000000           0.555000       0.750000         2.250000   
1759                   NaN                NaN            NaN              NaN   
1506            104.000000           0.590000       0.000000         0.000000   
2036             35.285714           0.414286       0.000000         0.142857   
2544             25.000000           0.610000       0.000000         0.000000   
4394                   NaN                NaN            NaN              NaN   
2954             26.000000           0.343333       0.333333         5.333333   
715                    NaN           0.200000       0.300000         1.350000   
2681             10.631579           0.508421       0.421053         0.315789   
4330             37.875000           0.367500       1.000000         1.125000   

      B_avg_TD_pct  B_longest_win_streak  B_losses  B_total_rounds_fought  \
1763      0.560000                     4         0                      7   
1759           NaN                     0         0                      0   
1506      0.000000                     1         0                      2   
2036      0.142857                     5         2                     14   
2544      0.000000                     3         1                      6   
4394           NaN                     0         0                      0   
2954      0.573333                     1         1                      9   
715       0.260000                     2         1                      7   
2681      0.144737                     3         6                     29   
4330     

In [16]:
# Dummy classifier used for simple model for comparison after optimization-DummyClassifier will allow us to run model with null values
simple_model = DummyClassifier(random_state=42)
simple_model.fit(X_train, y_train)

DummyClassifier(random_state=42)

In [17]:
prediction = simple_model.predict(X_test)
accuracy_score(y_test, prediction)

0.5175131348511384

#### Our simple model can predict winners with ~ 51.75% accuracy